src/rtc_stream.cpp

```diff
@@ -0,0 +1,418 @@
+
+void rtc_stream_impl::issue_read()
+{
+	TORRENT_ASSERT(m_read_handler);
+	TORRENT_ASSERT(m_read_buffer_size > 0);
+
+	if (!ensure_open()) return;
+
+	error_code ec;
+	std::size_t bytes_read = read_some(ec);
+	if (ec || bytes_read > 0) // error or immediate read
+	{
+		clear_read_buffers();
+		post(m_io_context, std::bind(std::exchange(m_read_handler, nullptr), ec, bytes_read));
+	}
+}
+
+void rtc_stream_impl::issue_write()
+{
+	TORRENT_ASSERT(m_write_handler);
+	TORRENT_ASSERT(m_write_buffer_size > 0);
+
+	if (!ensure_open()) return;
+
+	std::size_t const max_message_size = m_data_channel->maxMessageSize();
+	std::size_t bytes_written = 0;
+	bool is_buffered = false;
+	while (!m_write_buffer.empty())
+	{
+		bytes_written += write_data(max_message_size, is_buffered);
+	}
+
+	TORRENT_ASSERT(bytes_written == m_write_buffer_size);
+
+	if (!is_buffered)
+	{
+		m_write_buffer.clear();
+		m_write_buffer_size = 0;
+		post(m_io_context, std::bind(std::exchange(m_write_handler, nullptr), error_code{}, bytes_written));
+	}
+}
+
+std::size_t rtc_stream_impl::read_some(error_code& ec)
+{
+	if (!is_open())
+	{
+		ec = boost::asio::error::not_connected;
+		return 0;
+	}
+
+	std::size_t bytes_read = 0;
+
+	if (!m_incoming.empty())
+	{
+		std::size_t const copied = read_data(m_incoming.data(), m_incoming.size());
+		bytes_read += copied;
+		if (copied < m_incoming.size())
+		{
+			m_incoming.erase(m_incoming.begin(), m_incoming.begin() + std::ptrdiff_t(copied));
+			return bytes_read;
+		}
+
+		m_incoming.clear();
+	}
+
+	while (!m_read_buffer.empty() && m_incoming.empty() && !ec)
+	{
+		auto message = m_data_channel->receive();
+		if (!message) break;
+
+		std::visit(rtc::overloaded
+		{
+			[&](rtc::binary const& bin)
+			{
+				char const *data = reinterpret_cast<char const*>(bin.data());
+				std::size_t const size = bin.size();
+				std::size_t const copied = read_data(data, size);
+				bytes_read += copied;
+				if (copied < size)
+					m_incoming.assign(data + copied, data + size);
+			},
+			[&](rtc::string const&)
+			{
+				ec = errc::make_error_code(errc::bad_message);
+			}
+        }
+        , *message);
+	}
+
+	return bytes_read;
+}
+
+std::size_t rtc_stream_impl::write_some(error_code& ec)
+{
+	if (!is_open())
+	{
+		ec = boost::asio::error::not_connected;
+		return 0;
+	}
+
+	if (m_data_channel->bufferedAmount() > 0)
+	{
+		ec = boost::asio::error::would_block;
+		return 0;
+	}
+
+	std::size_t const max_message_size = m_data_channel->maxMessageSize();
+	std::size_t bytes_written = 0;
+	bool is_buffered = false;
+	while (!m_write_buffer.empty() && !is_buffered)
+	{
+		bytes_written += write_data(max_message_size, is_buffered);
+	}
+
+	return bytes_written;
+}
+
+void rtc_stream_impl::clear_read_buffers()
+{
+	m_read_buffer.clear();
+    m_read_buffer_size = 0;
+}
+
+void rtc_stream_impl::clear_write_buffers()
+{
+	m_write_buffer.clear();
+	m_write_buffer_size = 0;
+}
+
+std::size_t rtc_stream_impl::read_data(char const *data, std::size_t size)
```

Comment:
it seems it would be less error prone for this to take a `span<char const>`. The part "data += to_copy; size -= to_copy;" would just become `data = data.subspan(to_copy);`

Also, I got a bit confused about this function name at first. Would `incoming_data()` be a better name? (I thought it mean "I want to read, please fill in my buffer").

Reply:
Sounds good to me.